File: CTCI/Ch2/LinkedList.py
#Singly Linked Lists

# [] -> [] -> [] ->


class Node:
    def __init__(self, value,next_node = None):
        self.value = value
        self.next = next_node

    def __str__(self):
        return str(self.value)

class LinkedList:
    def __init__(self):
        self.head = None

    def __str__(self):
        str_ll = []
        curr = self.head

        while curr:
            str_ll.append(str(curr.value))
            curr = curr.next

        return ' -> '.join(str_ll)

    def is_empty(self):
        return self.head is None
    
    def add(self,value):
        if self.is_empty():
            self.head = Node(value)
        else:
            curr = self.head
            while curr.next:
                curr = curr.next
            curr.next = Node(value)
            return curr.next
    
    def group_add(self,values):
        for value in values:
            self.add(value)
    
    def to_list(self):
        if not self.head:
            return []
        
        curr = self.head
        arr = []

        while curr:
            arr.append(curr.value)
            curr = curr.next

        return arr
    
    def length(self):
        length = 0
        curr = self.head

        while curr:
            length += 1
            curr = curr.next

        return length
    
    def delete(self,key):
        if self.is_empty():
            return None
        
        if self.head.value == key:
            self.head = self.head.next
            return self.head.next
        
        temp = self.head

        while temp.next and temp.next.value != key:
            temp = temp.next
            
        if temp.next:
            temp.next = temp.next.next

        return self.head

File: Data_Structures_Java/LinkedList/LinkedList.java
package Data_Structures_Java.LinkedList;
//SLL
// While iterative traversal is more efficient due to recursion requiring a large call stack
// for bigger SLLs, both are used in the following class, in efforts to demonstrate knowledge of
// both.

public class LinkedList {
    Node head;

    LinkedList() {
        this.head = null;
    }

    public void add(int data) {
        if (isEmpty()) head = new Node(data, null);
        else add(head.next,data);
    }
    private void add(Node curr, int data) {
        if (curr.next == null) curr.next = new Node(data, null);
        else add(curr.next, data);
    }
    public void groupAdd(int[] arr) {
        for (int n: arr) add(n);
    }

    public boolean delete(int data) {
        if (head.data == data) {
            head = head.next;
            return true;
        }
        return delete(head.next, data);
    }
    private boolean delete(Node curr, int data) {
        if(curr == null || curr.next == null) return false;

        else if (curr.next.data == data) {
            curr.next = curr.next.next;
            return true;
        }
        
        return delete(curr.next, data);
    }

    public int[] toList() {
        if (isEmpty()) return new int[0];

        int size = getSize();
        int[] arr = new int[size];
        Node curr = head;

        for(int i = 0; i < size; i++) {
            arr[i] = curr.data;
            curr = curr.next;
        }
        return arr;
    }

    public int getSize() {
        if (isEmpty()) return 0;
        int size = 1;
        Node curr = head;

        while (curr != null) {
            size++;
            curr = curr.next;
        }

        return size;
    }

    public void print() {
        int[] ll = toList();

        for (int n: ll) System.out.print(n + "->");
    }

    public int pop() {
        if (head.next == null) {
            int popped = head.data;
            head = null;
            return popped;
        }
        Node curr = head;

        while(curr.next.next != null) {
            curr = curr.next;
        }

        int popped = curr.next.data;
        curr.next = null;

        return popped;
    }

    public boolean isEmpty() {
        if (head == null) return true;
        return false;
    }
}
